wavefront_pyformance/wavefront_reporter.py

```diff
@@ -27,26 +28,42 @@ def __init__(self, source='wavefront-pyformance', registry=None,
         self.prefix = prefix
         self.tags = tags or {}
 
+    @staticmethod
+    def decode_key(key):
+        """Decode encoded key into original key and dict of tags."""
+        if '-tags=' in key:
+            decoded_str = key.split('-tags=')
+            return decoded_str[0], json.loads(decoded_str[1])
+        return key, None
+
     def report_now(self, registry=None, timestamp=None):
         """Collect metrics from registry and report them to Wavefront."""
         registry = registry or self.registry
         metrics = registry.dump_metrics()
         for key in metrics.keys():
             is_delta = delta.is_delta_counter(key, registry)
             for value_key in metrics[key].keys():
+                metric_name, metric_tags = self.decode_key(key)
+                tags = self.tags
+                if metric_tags:
+                    tags = {}
+                    tags.update(self.tags)
```

Comment:
You already assigned `self.tags` to `tags` on line 47

Reply:
That's for metrics without tags in their name, it should be just `self.tags`. Otherwise, it should be `self.tags` plus tags decoded from its name.

Reply:
Correct me if I'm wrong, but having:

`tags = self.tags`

is the same as doing

```
tags = {}
tags.update(self.tags)
```
and it looks like latter is useless

Reply:
And, allocating a new dict and updating it with elements of another dict is way worse than just repointing the name to an existing one

Reply:
Unless you need to keep `self.tags` unchanged

Reply:
@ustinov yes self.tags should not be changed. It's the global tags applied to all metrics. We do the following to merge the global tags and the per metric level tags:

```
tags = {}
tags.update(self.tags)
tags.update(metric_tags)
```

Reply:
If you do, you may replace these with `tags = self.tags.copy()` in order to keep `self.tags` unaffected

Reply:
`self.tags.copy()` basically does the same thing, in one line